### PR TITLE
Loop Region Code Generation

### PR DESCRIPTION
**Pull request by phschaad**
This PR adapts code generation to make use of hierarchical control flow regions, and by extension `LoopRegion`s. This forms the fourth core element of the [plan to make loops first class citizens of SDFGs](https://github.com/orgs/spcl/projects/10) and marks the last element in the architecture.

By extending codegen with the capability of handling hierarchical control flow graphs and SDFGs, a myriad of complexities that come with control flow detection are circumvented, which currently lead to significant issues for certain SDFGs (e.g., #635 and #1586). Making use of control flow regions such as `LoopRegion`s instead allows codegen to be much less 'smart' and behave more akin to a lookup table that decides what code to generate for what SDFG element, making it significantly less error prone.

### CODE REVIEW

**Review by tbennun — CHANGES_REQUESTED** [review]

File: dace/codegen/control_flow.py [hunk 1]
```diff
@@ -454,19 +504,80 @@ def as_cpp(self, codegen, symbols) -> str:
...
+                expr += f'\n}} while({cond});\n'
+            else:
+                expr += f'for ({init}; {cond}; {update}) {{\n'
+                expr += _clean_loop_body(self.body.as_cpp(codegen, symbols))
+                expr += '\n}\n'
+        else:
+            if self.loop.inverted:
+                expr += 'do {\n'
+                expr += _clean_loop_body(self.body.as_cpp(codegen, symbols))
+                expr += f'\n}} while({cond});\n'
```
[the remark this answers is not in the document]
Comment (phschaad) [reply dace/codegen/control_flow.py:559]: yes, [`test_loop_do_for`](https://github.com/spcl/dace/blob/loop_region_codegen/tests/sdfg/loop_region_test.py#L89) tests for that.

File: dace/sdfg/analysis/schedule_tree/sdfg_to_tree.py [hunk 2]
```diff
@@ -726,8 +726,8 @@ def totree(node: cf.ControlFlow, parent: cf.GeneralBlock = None) -> List[tn.Sche
             # e.g., "SwitchCaseScope"
             raise tn.UnsupportedScopeException(type(node).__name__)
 
-        if node.first_state is not None:
-            result = [tn.StateLabel(state=node.first_state)] + result
+        if node.first_block is not None:
+            result = [tn.StateLabel(state=node.first_block)] + result
```
[the remark this answers is not in the document]
Comment (phschaad) [reply dace/sdfg/analysis/schedule_tree/sdfg_to_tree.py:732]: Added [tests for that](https://github.com/spcl/dace/blob/loop_region_codegen/tests/sdfg/loop_region_test.py#L199) now :-)

**Comment by phschaad** [comment]
@tbennun ping

File: dace/sdfg/analysis/schedule_tree/treenodes.py [hunk 3]
```diff
@@ -151,6 +151,38 @@ def as_string(self, indent: int = 0):
...
+    General loop scope (representing a loop region).
+    """
+    header: cf.GeneralLoopScope
+
+    def as_string(self, indent: int = 0):
+        loop = self.header.loop
+        if loop.update_statement and loop.init_statement and loop.loop_variable:
+            if loop.inverted:
+                pre_header = indent * INDENTATION + f'{loop.init_statement.as_string}\n'
+                header = indent * INDENTATION + 'do:\n'
```
Comment (tbennun) [inline dace/sdfg/analysis/schedule_tree/treenodes.py:166]: ha, nice. A good reminder that schedule trees are NOT Python :)

File: tests/sdfg/loop_region_test.py [hunk 4]
```diff
@@ -100,19 +83,12 @@ def test_loop_do_for():
...
-    a_validation = np.zeros([10], dtype=np.float32)
-    a_test = np.zeros([10], dtype=np.float32)
-    sdfg(A=a_test)
-    for i in range(10):
-        a_validation[i] = i
-    assert np.allclose(a_validation, a_test)
 
-
-def test_triple_nested_for():
+def _make_tripple_nested_for_loop() -> SDFG:
```
Comment (tbennun) [inline tests/sdfg/loop_region_test.py:89]:
```suggestion
def _make_triple_nested_for_loop() -> SDFG:
```

File: tests/sdfg/loop_region_test.py [hunk 5]
```diff
@@ -164,9 +196,79 @@ def test_triple_nested_for():
...
+
+    assert stree.as_string() == (f'{tn.INDENTATION}i = 0\n' +
+                                 f'{tn.INDENTATION}do:\n' +
+                                 f'{2 * tn.INDENTATION}A[i] = tasklet()\n' +
+                                 f'{2 * tn.INDENTATION}i = (i + 1)\n' +
+                                 f'{tn.INDENTATION}while (i < 10)')
+
+
+def test_loop_to_stree_triple_nested_for():
+    sdfg = _make_tripple_nested_for_loop()
```
Comment (tbennun) [inline tests/sdfg/loop_region_test.py:252]:
```suggestion
    sdfg = _make_triple_nested_for_loop()
```

File: tests/sdfg/loop_region_test.py [hunk 6]
```diff
@@ -146,6 +122,62 @@ def test_triple_nested_for():
...
+
+def test_loop_do_while():
+    sdfg = _make_do_while_loop()
+
+    assert sdfg.is_valid()
+
+    a_validation = np.zeros([11], dtype=np.float32)
+    a_test = np.zeros([11], dtype=np.float32)
+    a_validation[10] = 10
+    sdfg(A=a_test)
```
Comment (tbennun) [inline tests/sdfg/loop_region_test.py:162]: I think we want to test whether `generate_code()` has `do {` in it too.

**Review by tbennun — CHANGES_REQUESTED** [review]
Very very minor changes

**Review by tbennun — APPROVED** [review]

**Pull request merged**